File: R/R/io.R
#' @importFrom rlang abort
#' @importFrom arrow read_parquet write_parquet as_arrow_table
#' @importFrom tibble tibble as_tibble
#' @importFrom jsonlite toJSON fromJSON
NULL

SCHEMA_VERSION <- "1.0"
LIFECYCLE_VALS <- c("start", "complete", "schedule", "withdraw", "suspend", "resume")
REQUIRED_COLS  <- c("case_id", "activity", "timestamp")

# ---- validate_eventlog -------------------------------------------------------

#' Validate an event log against the processmine schema (v1).
#'
#' @param log A data frame / tibble.
#' @return `log` invisibly on success; throws on violation.
#' @export
validate_eventlog <- function(log) {
  missing_cols <- setdiff(REQUIRED_COLS, names(log))
  if (length(missing_cols) > 0) {
    rlang::abort(paste0("Missing required column(s): ", paste(missing_cols, collapse = ", ")))
  }

  for (col in REQUIRED_COLS) {
    if (anyNA(log[[col]])) {
      rlang::abort(paste0(
        "Column '", col, "' contains NA values; required columns must be non-null."
      ))
    }
  }

  .assert_utc(log$timestamp, "timestamp")

  if ("start_timestamp" %in% names(log)) {
    .assert_utc(log$start_timestamp, "start_timestamp", allow_na = TRUE)
    both <- !is.na(log$start_timestamp)
    if (any(log$timestamp[both] < log$start_timestamp[both])) {
      rlang::abort("start_timestamp must be <= timestamp for all events.")
    }
  }

  if ("lifecycle" %in% names(log)) {
    vals <- unique(stats::na.omit(log$lifecycle))
    bad  <- setdiff(vals, LIFECYCLE_VALS)
    if (length(bad) > 0) {
      rlang::abort(paste0(
        "Invalid lifecycle value(s): ", paste(bad, collapse = ", "),
        ". Allowed: ", paste(LIFECYCLE_VALS, collapse = ", ")
      ))
    }
  }

  invisible(log)
}

.assert_utc <- function(x, name, allow_na = FALSE) {
  if (!inherits(x, "POSIXct")) {
    rlang::abort(paste0("Column '", name, "' must be POSIXct."))
  }
  tz <- attr(x, "tzone")
  if (is.null(tz) || !(tz %in% c("UTC", "GMT"))) {
    rlang::abort(paste0("Column '", name, "' must have tz = UTC (got '", tz %||% "NULL", "')."))
  }
}

`%||%` <- function(a, b) if (!is.null(a)) a else b

# ---- read_xes ----------------------------------------------------------------

#' Read an XES event log file into a processmine tibble.
#'
#' @param path Path to a `.xes` file.
#' @return A validated tibble conforming to the processmine schema.
#' @export
read_xes <- function(path) {
  doc    <- xml2::read_xml(path)
  ns     <- c(xes = "http://www.xes-standard.org/")
  traces <- xml2::xml_find_all(doc, ".//xes:trace|.//trace", ns)

  rows <- vector("list", 0L)

  for (tr in traces) {
    case_id    <- .xes_string(tr, "concept:name")
    case_attrs <- .xes_extra_attrs(tr, c("concept:name"))

    events <- xml2::xml_find_all(tr, ".//xes:event|.//event", ns)
    for (ev in events) {
      activity    <- .xes_string(ev, "concept:name")
      ts          <- .xes_date(ev, "time:timestamp")
      resource    <- .xes_string(ev, "org:resource",          default = NA_character_)
      lifecycle   <- .xes_string(ev, "lifecycle:transition",  default = NA_character_)
      event_attrs <- .xes_extra_attrs(ev, c("concept:name", "time:timestamp",
                                            "org:resource", "lifecycle:transition"))
      rows[[length(rows) + 1L]] <- list(
        case_id     = case_id,
        activity    = activity,
        timestamp   = ts,
        resource    = resource,
        lifecycle   = lifecycle,
        case_attrs  = list(case_attrs),
        event_attrs = list(event_attrs)
      )
    }
  }

  log <- tibble::tibble(
    case_id     = vapply(rows, `[[`, character(1), "case_id"),
    activity    = vapply(rows, `[[`, character(1), "activity"),
    timestamp   = do.call(c, lapply(rows, `[[`, "timestamp")),
    resource    = vapply(rows, `[[`, character(1), "resource"),
    lifecycle   = vapply(rows, `[[`, character(1), "lifecycle"),
    case_attrs  = lapply(rows, function(r) r$case_attrs[[1]]),
    event_attrs = lapply(rows, function(r) r$event_attrs[[1]])
  )

  validate_eventlog(log)
  log
}

.xes_string <- function(node, key, default = NA_character_) {
  child <- xml2::xml_find_first(node, paste0(".//*[@key='", key, "']"))
  if (inherits(child, "xml_missing")) return(default)
  val <- xml2::xml_attr(child, "value")
  if (is.na(val)) default else val
}

.xes_date <- function(node, key) {
  raw <- .xes_string(node, key, default = NA_character_)
  if (is.na(raw)) return(as.POSIXct(NA, tz = "UTC"))
  as.POSIXct(raw, format = "%Y-%m-%dT%H:%M:%OSZ", tz = "UTC")
}

.xes_extra_attrs <- function(node, skip_keys) {
  children <- xml2::xml_children(node)
  out <- character(0)
  for (ch in children) {
    k <- xml2::xml_attr(ch, "key")
    v <- xml2::xml_attr(ch, "value")
    if (!is.na(k) && !(k %in% skip_keys) && !is.na(v) &&
        xml2::xml_name(ch) %in% c("string", "int", "float", "boolean")) {
      out[k] <- as.character(v)
    }
  }
  out
}

# ---- read_csv_eventlog -------------------------------------------------------

#' Read a CSV file into a processmine event log tibble.
#'
#' Maps arbitrary CSV column names to the processmine schema, parses
#' timestamps, and calls `validate_eventlog()` before returning.
#'
#' @param path Path to a `.csv` file.
#' @param case_col Name of the column containing the case identifier.
#' @param activity_col Name of the column containing the activity name.
#' @param timestamp_col Name of the column containing the event completion timestamp.
#' @param start_timestamp_col Optional name of the column containing the event
#'   start timestamp. `NULL` to omit.
#' @param resource_col Optional name of the resource column. `NULL` to omit.
#' @param lifecycle_col Optional name of the lifecycle column. `NULL` to omit.
#' @param case_attrs_cols Optional character vector of column names to fold
#'   into the `case_attrs` map. These columns are removed from the top-level
#'   output and stored as named character vectors per event.
#' @param event_attrs_cols Optional character vector of column names to fold
#'   into the `event_attrs` map.
#' @param timestamp_format `strptime`-style format string for parsing
#'   timestamps (e.g. `"%Y-%m-%d %H:%M:%S"`). `NULL` lets R guess.
#' @param tz Source timezone of the timestamp strings. Defaults to `"UTC"`.
#'   Set to the actual timezone if your CSV stores local times (e.g.
#'   `"America/New_York"`); they will be converted to UTC.
#' @param ... Additional arguments passed to [utils::read.csv()].
#'
#' @return A validated tibble conforming to the processmine schema.
#' @export
read_csv_eventlog <- function(
  path,
  case_col            = "case_id",
  activity_col        = "activity",
  timestamp_col       = "timestamp",
  start_timestamp_col = NULL,
  resource_col        = NULL,
  lifecycle_col       = NULL,
  case_attrs_cols     = NULL,
  event_attrs_cols    = NULL,
  timestamp_format    = NULL,
  tz                  = "UTC",
  ...
) {
  raw <- utils::read.csv(path, stringsAsFactors = FALSE, ...)

  # Check required source columns exist
  required_src <- c(case_col, activity_col, timestamp_col)
  optional_src <- c(start_timestamp_col, resource_col, lifecycle_col,
                    case_attrs_cols, event_attrs_cols)
  missing_src  <- setdiff(c(required_src, optional_src), names(raw))
  if (length(missing_src) > 0) {
    rlang::abort(paste0(
      "Column(s) not found in CSV: ", paste(missing_src, collapse = ", ")
    ))
  }

  # Build log with canonical names
  log <- tibble::tibble(
    case_id  = as.character(raw[[case_col]]),
    activity = as.character(raw[[activity_col]]),
    timestamp = .parse_csv_timestamp(raw[[timestamp_col]], timestamp_format, tz)
  )

  if (!is.null(start_timestamp_col)) {
    log$start_timestamp <- .parse_csv_timestamp(
      raw[[start_timestamp_col]], timestamp_format, tz
    )
  }
  if (!is.null(resource_col)) {
    log$resource <- as.character(raw[[resource_col]])
  }
  if (!is.null(lifecycle_col)) {
    log$lifecycle <- as.character(raw[[lifecycle_col]])
  }

  # Fold extra columns into attrs maps
  if (!is.null(case_attrs_cols)) {
    log$case_attrs <- lapply(seq_len(nrow(raw)), function(i) {
      vapply(case_attrs_cols, function(col) as.character(raw[[col]][i]), character(1))
    })
  }
  if (!is.null(event_attrs_cols)) {
    log$event_attrs <- lapply(seq_len(nrow(raw)), function(i) {
      vapply(event_attrs_cols, function(col) as.character(raw[[col]][i]), character(1))
    })
  }

  validate_eventlog(log)
  log
}

.parse_csv_timestamp <- function(x, fmt, tz) {
  parsed <- if (is.null(fmt)) {
    as.POSIXct(x, tz = tz)
  } else {
    as.POSIXct(x, format = fmt, tz = tz)
  }
  # Convert to UTC if source tz differs
  attr(parsed, "tzone") <- "UTC"
  parsed
}

# ---- write_eventlog_parquet --------------------------------------------------

#' Write a validated event log to a Parquet file.
#'
#' `case_attrs` and `event_attrs` list-columns are serialised to JSON strings
#' at the Parquet boundary for portability. Use `read_eventlog_parquet()` to
#' round-trip them back to named character vectors.
#'
#' @param log A data frame conforming to the processmine schema.
#' @param path Destination path for the `.parquet` file.
#' @return `path` invisibly.
#' @export
write_eventlog_parquet <- function(log, path) {
  validate_eventlog(log)
  df <- as.data.frame(log, stringsAsFactors = FALSE)

  # Ensure UTC tz attribute is preserved after coercion
  attr(df$timestamp, "tzone") <- "UTC"
  if ("start_timestamp" %in% names(df)) {
    attr(df$start_timestamp, "tzone") <- "UTC"
  }

  # Serialise map-like list columns to JSON strings
  if ("case_attrs" %in% names(df)) {
    df$case_attrs <- vapply(df$case_attrs, .attrs_to_json, character(1))
  }
  if ("event_attrs" %in% names(df)) {
    df$event_attrs <- vapply(df$event_attrs, .attrs_to_json, character(1))
  }

  tbl <- arrow::as_arrow_table(tibble::as_tibble(df))
  tbl <- tbl$ReplaceSchemaMetadata(list(schema_version = SCHEMA_VERSION))
  arrow::write_parquet(tbl, path)
  invisible(path)
}

.attrs_to_json <- function(x) {
  if (length(x) == 0) return("{}")
  # Accept both named lists and named character vectors
  x <- vapply(x, function(v) as.character(v)[1L], character(1L))
  as.character(jsonlite::toJSON(as.list(x), auto_unbox = TRUE))
}

# ---- read_eventlog_parquet ---------------------------------------------------

#' Read a Parquet event log written by processmine.
#'
#' @param path Path to a `.parquet` file.
#' @return A validated tibble conforming to the processmine schema.
#' @export
read_eventlog_parquet <- function(path) {
  tbl <- arrow::read_parquet(path, as_data_frame = FALSE)

  meta    <- tbl$schema$metadata
  version <- meta[["schema_version"]]
  if (is.null(version)) {
    rlang::abort("Parquet file is missing 'schema_version' metadata.")
  }
  major <- as.integer(strsplit(version, "\\.")[[1]][1])
  if (major != 1L) {
    rlang::abort(paste0(
      "Unsupported schema version '", version, "'. ",
      "Only major version 1 is supported."
    ))
  }

  log <- tibble::as_tibble(tbl)

  # Deserialise JSON string map columns back to named character vectors
  if ("case_attrs" %in% names(log)) {
    log$case_attrs <- lapply(log$case_attrs, .json_to_attrs)
  }
  if ("event_attrs" %in% names(log)) {
    log$event_attrs <- lapply(log$event_attrs, .json_to_attrs)
  }

  # Restore UTC tz attribute (Arrow round-trip may drop tzone on POSIXct)
  if ("timestamp" %in% names(log)) {
    attr(log$timestamp, "tzone") <- "UTC"
  }
  if ("start_timestamp" %in% names(log)) {
    attr(log$start_timestamp, "tzone") <- "UTC"
  }

  validate_eventlog(log)
  log
}

.json_to_attrs <- function(x) {
  if (is.null(x) || identical(x, "{}") || identical(x, "")) return(character(0))
  obj <- jsonlite::fromJSON(x, simplifyVector = TRUE)
  if (length(obj) == 0) return(character(0))
  vapply(obj, function(v) as.character(v)[1L], character(1L))
}

File: python/processmine_ml/io.py
"""processmine_ml/io.py — I/O for the processmine event-log schema (v1)."""

from __future__ import annotations

import json
import xml.etree.ElementTree as ET
from pathlib import Path
from typing import Any

import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq

SCHEMA_VERSION = "1.0"
LIFECYCLE_VALS = frozenset({"start", "complete", "schedule", "withdraw", "suspend", "resume"})
REQUIRED_COLS  = ("case_id", "activity", "timestamp")

_PARQUET_SCHEMA = pa.schema([
    pa.field("case_id",         pa.string(),                  nullable=False),
    pa.field("activity",        pa.string(),                  nullable=False),
    pa.field("timestamp",       pa.timestamp("us", tz="UTC"), nullable=False),
    pa.field("start_timestamp", pa.timestamp("us", tz="UTC"), nullable=True),
    pa.field("resource",        pa.string(),                  nullable=True),
    pa.field("lifecycle",       pa.string(),                  nullable=True),
    pa.field("case_attrs",      pa.string(),                  nullable=True),
    pa.field("event_attrs",     pa.string(),                  nullable=True),
]).with_metadata({"schema_version": SCHEMA_VERSION})


# ---- validate_eventlog -------------------------------------------------------

def validate_eventlog(df: pd.DataFrame) -> pd.DataFrame:
    """Validate *df* against the processmine schema v1.

    Returns *df* unchanged on success; raises ``ValueError`` on any violation.
    """
    missing = [c for c in REQUIRED_COLS if c not in df.columns]
    if missing:
        raise ValueError(f"Missing required column(s): {', '.join(missing)}")

    for col in REQUIRED_COLS:
        if df[col].isna().any():
            raise ValueError(
                f"Column '{col}' contains Null/NA values; required columns must be non-null."
            )

    _assert_utc(df["timestamp"], "timestamp")
    if "start_timestamp" in df.columns:
        _assert_utc(df["start_timestamp"], "start_timestamp", allow_na=True)
        both = df["start_timestamp"].notna() & df["timestamp"].notna()
        if (df.loc[both, "timestamp"] < df.loc[both, "start_timestamp"]).any():
            raise ValueError("start_timestamp must be <= timestamp for all events.")

    if "lifecycle" in df.columns:
        bad = set(df["lifecycle"].dropna().unique()) - LIFECYCLE_VALS
        if bad:
            raise ValueError(
                f"Invalid lifecycle value(s): {bad}. "
                f"Allowed: {sorted(LIFECYCLE_VALS)}"
            )

    return df


def _assert_utc(series: pd.Series, name: str, allow_na: bool = False) -> None:
    dtype = series.dtype
    if not hasattr(dtype, "tz"):
        raise ValueError(f"Column '{name}' must be a timezone-aware datetime (got {dtype}).")
    tz = str(dtype.tz)
    if tz not in ("UTC", "utc"):
        raise ValueError(f"Column '{name}' must have tz=UTC (got '{tz}').")


# ---- read_xes ----------------------------------------------------------------

_XES_NS         = "http://www.xes-standard.org/"
_STD_TRACE_KEYS = {"concept:name"}
_STD_EVENT_KEYS = {"concept:name", "time:timestamp", "org:resource", "lifecycle:transition"}
_STR_ELEM_TAGS  = {"string", "int", "float", "boolean"}


def read_xes(path: str | Path) -> pd.DataFrame:
    """Parse an XES event log file and return a validated DataFrame."""
    tree = ET.parse(str(path))
    root = tree.getroot()

    def _tag(elem: ET.Element) -> str:
        return elem.tag.replace(f"{{{_XES_NS}}}", "")

    rows: list[dict[str, Any]] = []

    for tr in root:
        if _tag(tr) != "trace":
            continue
        case_id, case_attrs = _parse_trace_attrs(tr, _tag)
        for ev in tr:
            if _tag(ev) != "event":
                continue
            row = _parse_event(ev, _tag)
            row["case_id"]    = case_id
            row["case_attrs"] = dict(case_attrs)
            rows.append(row)

    if not rows:
        return _empty_df()

    df = pd.DataFrame(rows)
    df["timestamp"] = (
        pd.to_datetime(df["timestamp"], utc=True, format="ISO8601").dt.as_unit("us")
    )
    if "start_timestamp" in df.columns:
        mask = df["start_timestamp"].notna()
        if mask.any():
            df.loc[mask, "start_timestamp"] = (
                pd.to_datetime(df.loc[mask, "start_timestamp"], utc=True, format="ISO8601")
                .dt.as_unit("us")
            )
        df["start_timestamp"] = df["start_timestamp"].astype("datetime64[us, UTC]")

    return validate_eventlog(df)


def _parse_trace_attrs(
    trace: ET.Element, tag_fn: Any
) -> tuple[str, dict[str, str]]:
    case_id = ""
    extra: dict[str, str] = {}
    for child in trace:
        t = tag_fn(child)
        k = child.get("key", "")
        v = child.get("value", "")
        if k == "concept:name":
            case_id = v
        elif t in _STR_ELEM_TAGS and k not in _STD_TRACE_KEYS:
            extra[k] = str(v)
    return case_id, extra


def _parse_event(event: ET.Element, tag_fn: Any) -> dict[str, Any]:
    row: dict[str, Any] = {
        "activity":    None,
        "timestamp":   None,
        "resource":    None,
        "lifecycle":   None,
        "event_attrs": {},
    }
    extra: dict[str, str] = {}
    for child in event:
        t = tag_fn(child)
        k = child.get("key", "")
        v = child.get("value", "")
        if k == "concept:name":
            row["activity"] = v
        elif k == "time:timestamp":
            row["timestamp"] = v
        elif k == "org:resource":
            row["resource"] = v
        elif k == "lifecycle:transition":
            row["lifecycle"] = v
        elif t in _STR_ELEM_TAGS and k not in _STD_EVENT_KEYS:
            extra[k] = str(v)
    row["event_attrs"] = extra
    return row


def _empty_df() -> pd.DataFrame:
    return pd.DataFrame(columns=list(REQUIRED_COLS))


# ---- write_eventlog_parquet --------------------------------------------------

def write_eventlog_parquet(df: pd.DataFrame, path: str | Path) -> None:
    """Validate *df* and write it to a Parquet file at *path*."""
    validate_eventlog(df)
    df = df.copy()

    df["timestamp"] = _to_utc_us(df["timestamp"])
    if "start_timestamp" in df.columns:
        df["start_timestamp"] = _to_utc_us(df["start_timestamp"])

    arrays: list[pa.Array] = []
    fields: list[pa.Field] = []

    for field in _PARQUET_SCHEMA:
        name = field.name
        if name not in df.columns:
            if field.nullable:
                arrays.append(pa.nulls(len(df), type=field.type))
            else:
                raise ValueError(f"Required column '{name}' is missing.")
        elif name in ("case_attrs", "event_attrs"):
            arrays.append(_dict_col_to_json_array(df[name]))
        elif pa.types.is_timestamp(field.type):
            arrays.append(pa.array(df[name], type=field.type))
        else:
            col = df[name].where(df[name].notna(), other=None)
            arrays.append(pa.array(col.tolist(), type=field.type))
        fields.append(field)

    schema = pa.schema(fields).with_metadata({"schema_version": SCHEMA_VERSION})
    table  = pa.table(dict(zip([f.name for f in fields], arrays)), schema=schema)
    pq.write_table(table, str(path))


def _to_utc_us(series: pd.Series) -> pd.Series:
    if not hasattr(series.dtype, "tz"):
        series = pd.to_datetime(series, utc=True)
    elif str(series.dtype.tz) != "UTC":
        series = series.dt.tz_convert("UTC")
    return series.dt.as_unit("us")


def _dict_col_to_json_array(series: pd.Series) -> pa.Array:
    """Serialise a column of dicts to JSON strings for Parquet storage."""
    out: list[str | None] = []
    for val in series:
        if val is None or (isinstance(val, float) and pd.isna(val)):
            out.append(None)
        else:
            out.append(json.dumps({str(k): str(v) for k, v in (val.items() if isinstance(val, dict) else val)}, sort_keys=True))
    return pa.array(out, type=pa.string())


# ---- read_eventlog_parquet ---------------------------------------------------

def read_eventlog_parquet(path: str | Path) -> pd.DataFrame:
    """Read a processmine Parquet file and return a validated DataFrame."""
    table = pq.read_table(str(path))

    meta    = table.schema.metadata or {}
    raw     = meta.get(b"schema_version") or meta.get("schema_version")
    version = raw.decode() if isinstance(raw, bytes) else (raw or "")
    if not version:
        raise ValueError("Parquet file is missing 'schema_version' metadata.")
    major = int(version.split(".")[0])
    if major != 1:
        raise ValueError(
            f"Unsupported schema version '{version}'. Only major version 1 is supported."
        )

    df = table.to_pandas(timestamp_as_object=False)

    for ts_col in ("timestamp", "start_timestamp"):
        if ts_col in df.columns and hasattr(df[ts_col].dtype, "tz"):
            df[ts_col] = df[ts_col].dt.as_unit("us").dt.tz_convert("UTC")

    # Deserialise JSON string attr columns back to Python dicts
    for attr_col in ("case_attrs", "event_attrs"):
        if attr_col in df.columns:
            df[attr_col] = df[attr_col].apply(
                lambda x: json.loads(x) if isinstance(x, str) else ({} if x is None else x)
            )

    return validate_eventlog(df)
